blockchain/consensus/deleg_cns/src/proposer.rs

```diff
@@ -0,0 +1,135 @@
+// finalized total order of transactions, which the validations
+// also access to check if the Filecoin blocks reflect the same.
+
+/// `DelegatedProposer` is a transient construct only created on the
+/// node doing all block proposals, it is responsible for doing the
+/// infinite loop of block creation. It needs access to the private
+/// key corresponding to the ID of the only actor allowed to sign
+/// blocks.
+pub struct DelegatedProposer {
+    miner_addr: Address,
+    key: Key,
+}
+
+impl DelegatedProposer {
+    pub(crate) fn new(miner_addr: Address, key: Key) -> Self {
+        Self { miner_addr, key }
+    }
+
+    async fn create_block<DB>(
+        &self,
+        mpool: &impl MessagePoolApi,
+        state_manager: &Arc<StateManager<DB>>,
+        base: &Arc<Tipset>,
+    ) -> anyhow::Result<GossipBlock>
+    where
+        DB: BlockStore + Sync + Send + 'static,
+    {
+        let block_delay = state_manager.chain_config().block_delay_secs;
+        let smoke_height = state_manager.chain_config().epoch(Height::Smoke);
+
+        let (parent_state_root, parent_receipts) = state_manager.tipset_state(base).await?;
+        let parent_base_fee =
+            chain::compute_base_fee(state_manager.blockstore(), base, smoke_height)?;
+
+        let parent_weight = DelegatedConsensus::weight(state_manager.blockstore(), base)?;
+        let msgs = mpool.select_signed(state_manager, base).await?;
+        let msgs = msgs.iter().map(|m| m.as_ref()).collect();
+        let persisted = chain::persist_block_messages(state_manager.blockstore(), msgs)?;
+
+        let mut header = BlockHeader::builder()
+            .messages(persisted.msg_cid)
+            .bls_aggregate(Some(persisted.bls_agg))
+            .miner_address(self.miner_addr)
+            .weight(parent_weight)
+            .parent_base_fee(parent_base_fee)
+            .parents(base.key().clone())
+            .epoch(base.epoch() + 1)
+            .timestamp(base.min_timestamp() + block_delay)
+            .state_root(parent_state_root)
+            .message_receipts(parent_receipts)
+            .build()?;
+
+        let sig = key_management::sign(
+            *self.key.key_info.key_type(),
+            self.key.key_info.private_key(),
+            &header.to_signing_bytes(),
+        )?;
+
+        header.signature = Some(sig);
+
+        Ok(GossipBlock {
+            header,
+            bls_messages: persisted.bls_cids,
+            secpk_messages: persisted.secp_cids,
+        })
+    }
+}
+
+#[async_trait]
+impl Proposer for DelegatedProposer {
+    async fn run<DB, MP>(
+        self,
+        state_manager: Arc<StateManager<DB>>,
+        mpool: &MP,
+        submitter: &SyncGossipSubmitter,
+    ) -> anyhow::Result<()>
+    where
+        DB: BlockStore + Sync + Send + 'static,
+        MP: MessagePoolApi + Send + Sync + 'static,
+    {
+        // TODO: Ideally these should not be coming through the `StateManager`.
```

Comment:
Yes, I think we should only give it limited capabilities. The `MessagePool` has too many public mutable fields, so that was a non-starter to share (anyone with a direct reference to the message pool could insert/remove messages from its caches at will). The `StateManager` itself is not too bad, but the `ChainStore` itself has public methods like `set_genesis` and `set_heaviest_tipset` that we should not be able to call without going through the proper channels. 

The blockstore is another one that would be best if it was passed directly, if access is needed. Currently you can reach it via `ChainStore.db`, `ChainStore.blockstore()`, `ChainStore.blockstore_cloned()`, `StateManager.blockstore()`, `StateManger.blockstore_cloned()`, `StateManager.chain_store().blockstore()`, etc.

The `Provider` that the `MessagePool` gets is better at limiting access to what the mempool can use, but in return exposed via the public `api` field, rather than kept for itself, so again it's not perfect. Also, at some point interfaces like this tend to become annoying as well: they collect all dependencies a component needs into one mockable package, but they also make a mess of individual concepts, like having a state manager, a chain store and a block store as _things_. It's tempting to put too much logic into the methods of the provider, rather than what it provides to. 

It is a balancing act. Perhaps separating interfaces to a reader/writer part is a good start.

Reply:
That's a great point. I think is definitely worth a refactor as we keep building the protocol.